src/azul/indexer/__init__.py

```diff
@@ -59,6 +68,101 @@ class BundleFQID(SupportsLessThan):
+        if self.partition:
+            assert isinstance(self.partition, int), self.partition
+            partition = self.partition
+        else:
+            partition = config.partition_prefix_length
+        partition_prefix = first(self._partition_generator(partition))
+        validate_uuid_prefix(self.common + partition_prefix)
+
+    @classmethod
+    def parse(cls, prefix: str) -> 'Prefix':
+        """
+        >>> Prefix.parse('aa/1')
+        Prefix(common='aa', partition=1)
+
+        >>> p = Prefix.parse('aa')
+        >>> p
+        Prefix(common='aa', partition=None)
+        >>> str(p)
+        'aa'
+        >>> from unittest.mock import patch
+        >>> import os
+        >>> with patch.dict(os.environ, AZUL_PARTITION_PREFIX_LENGTH='2'):
+        ...     p.effective
+        Prefix(common='aa', partition=2)
+        >>> Prefix.parse('aa/')
+        Traceback (most recent call last):
+        ...
+        azul.RequirementError: Prefix source: `aa/` cannot end in a `/`.
+        >>> Prefix.parse('8f538f53/1')
+        Traceback (most recent call last):
+        ...
+        azul.uuids.InvalidUUIDPrefixError: '8f538f530' is not a valid UUID prefix.
+        """
+        source_delimiter = '/'
+        reject(prefix.endswith(source_delimiter),
+               f'Prefix source: `{prefix}` cannot end in a `{source_delimiter}`.')
+        if prefix == '':
+            entry = ''
+            partition = None
+        else:
+            try:
+                entry, partition = prefix.split(source_delimiter)
+            except ValueError:
+                entry = prefix
+                partition = None
+            if partition:
+                try:
+                    partition = int(partition)
+                except ValueError:
+                    raise ValueError('Partition prefix length must be an integer: ',
+                                     partition)
+        validate_uuid_prefix(entry)
+        return cls(common=entry, partition=partition)
+
+    @cached_property
+    def effective(self) -> 'Prefix':
+        if self.partition is None:
+            return attr.evolve(self, partition=config.partition_prefix_length)
+        else:
+            return self
+
+    def partition_prefixes(self) -> Iterator[str]:
+        """
+        >>> list(Prefix.parse('/0').partition_prefixes())
+        ['']
+        >>> sorted(Prefix.parse('/1').partition_prefixes())
+        ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f']
+        >>> len(set(Prefix.parse('/2').partition_prefixes()))
+        256
+        """
+        for partition_prefix in self._partition_generator(self.partition):
+            validate_uuid_prefix(self.common + partition_prefix)
+            yield partition_prefix
+
+    def _partition_generator(self, partition) -> Iterator[str]:
+        return map(''.join, product('0123456789abcdef', repeat=partition))
+
+    def __str__(self):
```

Comment:
Missing doctest for https://github.com/DataBiosphere/azul/pull/2931#discussion_r674388377